cmd/otel-allocator/allocation/allocator.go

```diff
@@ -88,119 +88,150 @@ func (allocator *Allocator) Collectors() map[string]*collector {
+	defer timer.ObserveDuration()
 
 	allocator.m.Lock()
 	defer allocator.m.Unlock()
+	newCollectors, removedCollectors := allocator.getCollectorChanges(collectors)
 	if len(collectors) == 0 {
 		log.Info("No collector instances present")
 		return
+	} else if len(newCollectors) == 0 && len(removedCollectors) == 0 {
+		log.Info("No changes to the collectors found")
+		return
 	}
-	for k := range allocator.collectors {
+
+	// Clear existing collectors
+	for _, k := range removedCollectors {
 		delete(allocator.collectors, k)
+		targetsPerCollector.WithLabelValues(k).Set(0)
 	}
-
-	for _, i := range collectors {
+	// Insert the new collectors
+	for _, i := range newCollectors {
 		allocator.collectors[i] = &collector{Name: i, NumTargets: 0}
 	}
-	collectorsAllocatable.Set(float64(len(collectors)))
-}
-
-// AllocateTargets removes outdated targets and adds new ones from
-// waitingTargets. This method needs to be called to process the new target
-// updates. Until it is called, old targets will be served.
-func (allocator *Allocator) AllocateTargets() {
-	allocator.m.Lock()
-	timer := prometheus.NewTimer(timeToAssign.WithLabelValues("AllocateTargets"))
-	defer timer.ObserveDuration()
-	defer allocator.m.Unlock()
-	allocator.removeOutdatedTargets()
-	allocator.processWaitingTargets()
-}
-
-// ReallocateCollectors reallocates the targets among the new collector instances.
-func (allocator *Allocator) ReallocateCollectors() {
-	allocator.m.Lock()
-	timer := prometheus.NewTimer(timeToAssign.WithLabelValues("ReallocateCollectors"))
-	defer timer.ObserveDuration()
-	defer allocator.m.Unlock()
-	allocator.targetItems = make(map[string]*TargetItem)
-	allocator.processWaitingTargets()
-}
-
-// removeOutdatedTargets removes targets that are no longer available. This
-// method is called after a lock has been acquired in ReallocateCollectors().
-func (allocator *Allocator) removeOutdatedTargets() {
-	for k := range allocator.targetItems {
-		if _, ok := allocator.targetsWaiting[k]; !ok {
-			allocator.collectors[allocator.targetItems[k].Collector.Name].NumTargets--
-			delete(allocator.targetItems, k)
-		}
-	}
-}
 
-// processWaitingTargets processes the newly set targets. This method is called
-// after a lock has been acquired in AllocateTargets() or ReallocateCollectors().
-func (allocator *Allocator) processWaitingTargets() {
-	for k, v := range allocator.targetsWaiting {
-		if _, ok := allocator.targetItems[k]; !ok {
-			col := allocator.findNextCollector()
-			allocator.targetItems[k] = &v
-			targetItem := TargetItem{
-				JobName:   v.JobName,
-				Link:      LinkJSON{fmt.Sprintf("/jobs/%s/targets", url.QueryEscape(v.JobName))},
-				TargetURL: v.TargetURL,
-				Label:     v.Label,
-				Collector: col,
+	// find targets which need to be redistributed
+	var redistribute []*TargetItem
+	for _, item := range allocator.targetItems {
+		for _, s := range removedCollectors {
+			if item.Collector.Name == s {
+				redistribute = append(redistribute, item)
```

Comment:
```suggestion
				redistribute = append(redistribute, item)
				break
```

Maybe a micro-optimization, but no need to check any further collectors once you've found a match.

Reply:
yep, but good practice to do :)